### DISCUSSION TITLE
Patch operation - mapper output sets fields as undefined where not present in source

### DISCUSSION THREAD

**Discussion by jbeckton** [nartc/mapper#412] [General]
This issue seems out of the blue, I remember this working properly after updating to v7 but now all the sudden mapped object contains fields with a value of undefined where they did not exist in the source object.

When I send a patch operation to my api the dto may only contain the fields I want to update, the mapped output contains the missing fields with values set to undefined. This is causing the collection document fields to be changed to null which is very bad since the data is now lost.

I am hoping that I made a mistake somewhere but I don't remember having to do anything after the v7 install to get the missing source fields to be ignored in the destination. This can be a huge problem in production causing data loss, I need to figure out if the mapper lib just doesn't work this way anymore or that I inadvertently changed something to cause this behavior. 

`
SectionDto {
feature: 'listings',
 slug: 'hello-slug'
}
`

`
SectionMongoDocument {
  feature: 'listings',
  slug: 'hello-slug',
  name: undefined,
  order: undefined,
  published: undefined,
  description: undefined,
  icon: undefined,
  categories: undefined
}
`

**Comment by nartc** [comment]
That definitely sounds like a regression. Would you mind providing a quick simple reproduce so I can take a closer look?

**Comment by jbeckton** [comment]
````
AutomapperModule.forRoot({
      options: [{ name: 'sections', pluginInitializer: classes, errorHandler: new MapperErrorHandler() }],
      singular: false
    }),
````


````
export class SectionMongoDto {
  id?: string;
  @AutoMap()
  feature?: string;
  @AutoMap()
  slug?: string;
  @AutoMap()
  name?: string;
  @AutoMap()
  order?: number;
  @AutoMap()
  published?: boolean;
  @AutoMap()
  description?: string;
  @AutoMap()
  icon?: string;
  @AutoMap({ typeFn: () => CategoryMongoDto })
  categories?: CategoryMongoDto[];
}
````

````
export class SectionMongoDocument {
  _id?: ObjectId;
  __v?: number;
  @AutoMap()
  feature?: string;
  @AutoMap()
  slug?: string;
  @AutoMap()
  name?: string;
  @AutoMap()
  order?: number;
  @AutoMap()
  published?: boolean;
  @AutoMap()
  description?: string;
  @AutoMap()
  icon?: string;
  @AutoMap({ typeFn: () => CategoryMongoDocument })
  categories?: CategoryMongoDocument[];
}
````

````
mapper.createMap(SectionMongoDto, SectionMongoDocument).forMember((d) => d.categories, ignore());
````
````
const patchSectionDto = {
    "feature": "listings",
    "slug": "my-slug"
}
const sectionDocument = this.sectionMapper.map(patchSectionDto, SectionMongoDocument, SectionMongoDto);
````

**Comment by jbeckton** [comment]
I am currently using 7.2.1 but I upgraded when I noticed this issue, I was on ^7.0.2 so it could have started with one of the minor version updates. I will gladly provide more info upon request, I enjoyed using your library as it saved me a lot of time and code. I can't risk the data loss due to regression bugs so i'm gonna switch to Morphism. Best of luck.

**Reply by nartc** [reply]
@jbeckton Thanks again for the kind words. I've just had time to look into this and I can confirm that this is **NOT** a regression. I've tested with v5 and v6 of `@automapper/core` and this HAS ALWAYS BEEN the correct behavior (you can see here: https://stackblitz.com/edit/typescript-m5avgk?file=index.ts)

I think it might have something to do with either Mongo Driver or Mongoose (if you're using `mongoose`) instead of AutoMapper.

That said, have you tried doing this instead?

```ts
const section = await Section.findById(theId);
mapper.map(dto, SectionMongoDocument, SectionMongoDto, section); // map mutate

// section has been mutated with data from dto. Save it
await section.save();
```

Not entirely sure if it works with `mongoose`. I also found this (vaguely remember it was similar), maybe related: https://mongoosejs.com/docs/migrating_to_6.html#removed-omitundefined

**Reply by jbeckton** [reply]
@nartc  I am using the Mongo driver for Node. I'm actually porting from Mongoose to the Mongo driver. 

The snippet I included in my first post was from a console.log() statement before the database code. so the properties are being set with undefined straight from the mapper.

````
SectionMongoDocument { feature: 'listings', slug: 'hello-slug', name: undefined, order: undefined, published: undefined, description: undefined, icon: undefined, categories: undefined }
````

is / was there a config setting that gets passed to the mapper profile or mapper module that influences this behavior?

**Reply by nartc** [reply]
@jbeckton Yes, that is the expected behavior since the beginning of AutoMapper. Properties that aren't mapped (or aren't existing in the `sourceObj`) will become undefined. This is to maintain the shape of the model.

I meant that before in Mongoose, when you update an object, `undefined` properties might be stripped away for you. Eg:

```ts
const dto = {feature: 'listings', slug: 'hello-slug'};

// SectionMongoDocument { feature: 'listings', slug: 'hello-slug', name: undefined, order: undefined, published: undefined, description: undefined, icon: undefined, categories: undefined }
const doc = mapper.map(dto, SectionMongoDocument, SectionMongoDto);

// Pseudo code here: this update call might strip out "undefined" properties automatically.
Section.update(doc); //
```

You're saying that you are migrating off of Mongoose to vanilla Mongo Driver so that might be the case. I just want to confirm that this was not a regression with AutoMapper. If things behave differently for you, then it might be something else rather than AutoMapper itself.

From AutoMapper side, you can have 2 approaches:
- Use `afterMap()`. You can create a utility function called `stripUndefined()` and attach that function to `afterMap()` callback when you call `createMap()`  (https://automapperts.netlify.app/docs/misc/callbacks)
- Extend the `classes` plugin. Example here https://automapperts.netlify.app/docs/plugins-system/extend-plugin. In this example, I extend `classes` plugin to convert `null` to `undefined` but you can adjust it to strip `undefined` values instead.

**Reply by jbeckton** [reply]
Ok makes sense, as I removed the Mongoose implementation I started seeing this different behavior, I guess I did not realize that Mongoose was dealing with the undefined properties.

Thanks!

**Reply by jbeckton** [reply]
@nartc the afterMap() does not influence the destination object, I am still seeing properties with undefined values in the mapped result.

i'm not sure which of these is correct but neither produce the desired result.

````
const categoryDocument = this.sectionMapper.map<CategoryMongoDto, CategoryMongoDocument>(
      patchCategoryDto,
      CategoryMongoDocument,
      CategoryMongoDto,
      {
        afterMap: (source, destination) => {
            destination = removeEmpty(destination);
        }
      }
    );
```` 

````
const categoryDocument = this.sectionMapper.map<CategoryMongoDto, CategoryMongoDocument>(
      patchCategoryDto,
      CategoryMongoDocument,
      CategoryMongoDto,
      {
        afterMap: (source, destination) => {
          return removeEmpty(destination);
        }
      }
    );
````

**Reply by nartc** [reply]
What is `removeEmpty`?

**Reply by jbeckton** [reply]
removeEmpty is a function I wrote to strip properties from the object that contain undefined values

````
export const removeEmpty = (obj) => {
  const newObj = {};
  Object.keys(obj).forEach((key) => {
    if (obj[key] === Object(obj[key])) newObj[key] = removeEmpty(obj[key]);
    else if (obj[key] !== undefined) newObj[key] = obj[key];
  });
  return newObj;
};
````

**Reply by nartc** [reply]
@jbeckton `afterMap()` uses the references to the `source` and `destination`. So you can mutate the `destination` directly

```ts
const categoryDocument = this.sectionMapper.map<CategoryMongoDto, CategoryMongoDocument>(
      patchCategoryDto,
      CategoryMongoDocument,
      CategoryMongoDto,
      {
        afterMap: (_, destination) => {
            for (const key of Object.keys(destination)) {
               if (destination[key] === undefined) {
                  delete destination[key];
               }
            }
        }
      }
    );
```

this should work

**Reply by jbeckton** [reply]
@nartc so wouldn't just reassigning it work then?

````
const categoryDocument = this.sectionMapper.map<CategoryMongoDto, CategoryMongoDocument>(
      patchCategoryDto,
      CategoryMongoDocument,
      CategoryMongoDto,
      {
        afterMap: (source, destination) => {
            destination = removeEmpty(destination);
        }
      }
    );
````

**Reply by jbeckton** [reply]
I guess reassignment does not work... Here is a function that supports recursion for nested objects and it works as expected.

````
export const removeUndefinedFromRef = (obj) => {
  Object.keys(obj).forEach((key) => {
    if (obj[key] === Object(obj[key])) {
      removeUndefinedFromRef(obj[key]);
    } else if (obj[key] === undefined) {
      delete obj[key];
    }
  });
};
````

**Reply by nartc** [reply]
@jbeckton no because `removeEmpty()` returns a new reference. Pointing `destination` to a new reference doesn't work in the case of `afterMap()` here.

Yeah. That should work. You can attach that on `createMap()`

```ts
mapper.createMap(CaterogyMongoDto, CategoryMongoDocument)
   .afterMap(removeUndefinedFromRef);
```

This will run on every map you have for `mapper.map(dto, CategoryMongoDocument, CategoryMongoDto)` without having to attach `afterMap()` on the `.map()` operation. Of course, attaching the callback on `.map()` gives you better control and I guess easier to debug :)
